alidns/client.go

```diff
@@ -0,0 +1,171 @@
+package alidns
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+)
+
+// Client is an abstration of AliClint
+type Client struct {
+	Clint *AliClint
+	mutex sync.Mutex
+}
+
+func (p *Provider) getClient() error {
+	cred := newCredInfo(p.AccKeyID, p.AccKeySecret, p.RegionID)
+	return p.getAliClient(cred)
+}
+
+func (p *Provider) addDomainRecord(ctx context.Context, rc aliDomaRecord) (recID string, err error) {
+	p.mutex.Lock()
+	p.getClient()
+	p.Clint.addReqBody("Action", "AddDomainRecord")
+	p.Clint.addReqBody("DomainName", rc.DName)
+	p.Clint.addReqBody("RR", rc.Rr)
+	p.Clint.addReqBody("Type", rc.DTyp)
+	p.Clint.addReqBody("Value", rc.DVal)
+	p.Clint.addReqBody("TTL", fmt.Sprintf("%d", rc.TTL))
+	rs := aliResult{}
+	err = p.doAPIRequest(ctx, &rs)
+	recID = rs.RecID
+	p.mutex.Unlock()
```

Comment:
Though this appears to be technically safe right now, prefer using `defer` immediately after calling `Lock()`, that will be safer and easier to maintain going forward.

And similar for all other uses of Lock and Unlock.